Writing Slack bot using mongodb post (#143)
unicorn-utterances/unicorn-utterances: content/blog/making-a-slack-bot-with-node-and-mongo/index.md

---
{
	title: "Making a Slack Bot using NodeJS and MongoDB",
	description: 'Have you ever wanted to run native Java and Kotlin code from your mobile game written in Unity? Well you can! This article outlines how to set that up!',
	published: '2020-02-04T05:12:03.284Z',
	authors: ['crutchcorn'],
	tags: ['mongodb', 'node', 'slack'],
	attached: [],
	license: 'cc-by-nc-sa-4'
}

---

> Modern-day remote live communication has never been as efficient or fast as it is today. You're able to use services like Slack to join huge multi-channel communication workspaces for pleasure or buisness. These channels are often able to be super-powered by in-chat bots and applications that can inform you of new information from external services or even add new functionality to the chat. Luckily for us, Slack has put a lot of effort into making these extensions to Slack easy to write.

Reviewer: ```suggestion
Modern-day remote live communication has never been as efficient or fast as it is today. Services like Slack make it easy to join huge multi-channel communication workspaces for pleasure or business. These channels are often able to be super-powered by in-chat bots and applications that can inform you of new information from external services or even add new functionality to the chat. Luckily for us, Slack has put a lot of effort into making these extensions to Slack easy to write.
```

> One way they've eased the effort of their creation is by providing an SDK for Node developers to take and create extensions with. This post will outline how we can create a Slack bot to add functionality to chats.

Reviewer: ```suggestion
One way they've made extension development easier is by providing an SDK for Node developers to use and create extensions with. This post will outline how we can create a Slack bot to add functionality to chats.
```

# Initial Signup {#signup-for-dev-account}

> To start, we'll need to [signup for a developer account and create an app to host our applicaiton logic using this link](https://api.slack.com/apps). This will allow us to create new Slack apps and bots to add into our workspace.

Reviewer: ```suggestion
To start, we'll need to [sign up for a developer account and create an app to host our application logic using this link](https://api.slack.com/apps). This will allow us to create new Slack apps and bots to add into our workspace.
```

> ![The create app dialog once pressed "create app"](./create_app_dialog.png)

Reviewer: I'm not quite sure what the alt text is trying to say.

Author: ```suggestion
![The create app dialog that shows up once "create app" is pressed](./create_app_dialog.png)
```

> Enter in an app name, and assign the workspace you want the app to live during development. Once done, you should be greeted by a dashboard for your Slack app. You'll want to keep this screen open during development, as we'll be refering to it throughout this post. 

Reviewer: ```suggestion
Enter in an app name, and assign the workspace where you want the app to live during development. Once done, you should be greeted by a dashboard for your Slack app. You'll want to keep this screen open during development, as we'll be referring to it throughout this post.
```

![The initial screen that'll be shown when a new app is created](./initial_screen.png)

> This screen (and the tabs off to the side) are able to configure all of the interactions with Slack that we'll build upon with our code. We're even able to customize the look of our application in our Slack settings at the bottom of this homepage.

Reviewer: ```suggestion
This screen (and the tabs off to the side) provides the configuration for all of the interactions with Slack that we'll build upon with our code. We're even able to customize the look of our application in our Slack settings at the bottom of this homepage.
```

> ![Towards the bottom of the intial page will show how to customize the description and such](./display_info.png)

Reviewer: ```suggestion
![Towards the bottom of the initial page will show how to customize the description and such.](./display_info.png)
```

As mentioned previously, Slack provides an SDK for Node applications. [You can find the homepage for the npm package at the following URL.](https://github.com/slackapi/node-slack-sdk)

In order to quickly set up the SDK, we'll create a new directory for our code to live. Once we have a clear directory, we can run:

```
npm init -y
```

To setup an initial `package.json`. Once we have a `package.json`, we can add the packages we require to use the Slack SDK:

```
npm install @slack/web-api @slack/events-api
```

> After this, we'll then be able to use their example API from the README of their project as a baseline to our app:

Reviewer: ```suggestion
After this, we'll then be able to use their example API from the README of their project as a starter for our app:
```

```javascript
// index.js
// Initialize using signing secret from environment variables
const { createEventAdapter } = require('@slack/events-api');
// Slack requires a secret key to run your bot code. We'll find and figure out this signing secret thing in the next steps
const slackEvents = createEventAdapter(process.env.SLACK_SIGNING_SECRET);
const port = process.env.PORT || 3000;

// Attach listeners to events by Slack Event "type". See: https://api.slack.com/events/message.im
slackEvents.on('message', (event) => {
  console.log(`Received a message event: user ${event.user} in channel ${event.channel} says ${event.text}`);
});

// Handle errors (see `errorCodes` export)
slackEvents.on('error', console.error);

// Start a basic HTTP server
slackEvents.start(port).then(() => {
  // Listening on path '/slack/events' by default
  console.log(`server listening on port ${port}`);
});
```

> This code is what we'll need to run a `console.log` every time a user sends a message. However, _we'll need to setup more to get this code actually running due to Slack's permissions systems_ and such. For now, we'll save this code to `index.js` in the same folder we saved our `package.json` file.

Reviewer: ```suggestion
This code is what we'll need to run a `console.log` every time a user sends a message. However, _we'll need to set things up more to get this code actually working due to Slack's permissions systems_ and such. For now, we'll save this code to `index.js` in the same folder we saved our `package.json` file.
```

> Another thing that was mentioned in the code sample was the `process.env.SLACK_SIGNING_SECRET`. This is the key that Slack will use to connect your code to your workspace. We'll want to keep in mind how to store the signing secret (as the name implies, _we want to keep this key a secret as otherwise anyone can highjack your Slack app_). As the above code hints at, it's suggested to use an environmental file or configuration.

Reviewer: ```suggestion
Another thing that was mentioned in the code sample was the `process.env.SLACK_SIGNING_SECRET`. This is the key that Slack will use to connect your code to your workspace. We'll want to keep in mind how to store the signing secret (as the name implies, _we want to keep this key a secret as otherwise anyone can hijack your Slack app_). As the above code hints at, it's suggested to use an environment variable file or configuration.
```

> While environmental variables are typically assigned by system configurations, to make development easier, we'll setup a `.env` file with the expected credentials. Then, to inject the `.env` file contents into our `process`, we'll run our code using [the `env-cmd` package](https://www.npmjs.com/package/env-cmd). We'll start by installing the package:

Reviewer: ```suggestion
While environment variables are typically assigned by system configurations, we'll make development easier by setting up a `.env` file with the expected credentials. Then, to inject the `.env` file contents into our process, we'll run our code using [the `env-cmd` package](https://www.npmjs.com/package/env-cmd). We'll start by installing the package:
```

```
npm i env-cmd
```

This package will look for a `.env` file and inject it into your command that follows `env-cmd`. So, for example, you can **make a new file called `.env` and place the following contents in it**:

```
SLACK_SIGNING_SECRET=<SIGNING_SECRET_FROM_HOMESCREEN>
```

> Then, in your `package.json`, you can **edit your `start` command** to reflect the following:

Reviewer: ```suggestion
Then, in your `package.json`, you can **edit your `start` script** to reflect the following:
```

```json
{
  "scripts": {
    "start": "env-cmd node ./index.js"
  }
}
```

> Now, whenever your code shows `process.env.SLACK_SIGNING_SECRET`, it'll represent the value present in your `.env` file

Reviewer: ```suggestion
Now, whenever your code uses `process.env.SLACK_SIGNING_SECRET`, it'll represent the value present in your `.env` file.
```

# Development Hosting {#development-environment-setup}

In order to have these events called, we'll need to get a public URL to route to our local development server. In order to do this, we can [use `ngrok`](https://github.com/inconshreveable/ngrok) to host a public URL in our local environment:

```
npm i -D ngrok

> npx ngrok http 3000

Reviewer: is the `npm install` necessary here? npx will run the package regardless of whether it's installed.

Author: It will, but I'd rather leave the instruction in than leave it out. Npx is great but having to grab from the network every time is irksome and prevents me from adding `ngrok` to `script` files.

```

> Keep in mind that this should NOT be used to host your Slack application when you're ready to publish.

> > This should only be used during development process. In order to see how to deploy, you'll want to checkout [the section on doing so using Heroku](#deployment)

Reviewer: ```suggestion
> This should only be used during development process. In order to see how to deploy, you'll want to check out [the section on doing so using Heroku](#deployment).
```

> After doing so, you should be given an `ngrok.io` subdomain to map to your local IP address with a message like the following:

Reviewer: ```suggestion
After doing so, you should be given an `ngrok.io` subdomain to map to your local IP address with a message like this:
```

```
Forwarding https://9fca9f3e.ngrok.io -> http://localhost:3000
```

![Showing ngrok running in the terminal](./ngrok_running.png)

> We're now able to use this URL as a map to the external world to the local environment we're in. This is how we'll tell Slack to run our `index.js` file when we recieve a new event.

Reviewer: ```suggestion
We're now able to use this URL as a bridge between the external internet and the local environment we're in. This is how we'll tell Slack to run our `index.js` file when we receive a new event.
```

![Adding the ngrok domain into the "event subscription" area](./event_subscription_enable.png)

This should show "Verified" to explain that your domain is verified to have belonged to you, but the domain isn't saved yet; We first need to **add workspace events to subscribe to**. This is to ensure that any app doesn't simply have root permissions to everything for privacy and security's sake and instead has to ask for grainular permissions.

![Searching for oauth permissions to add to the event handler](./searching_events.png)

Let's say we want to handle all of the public messages to a channel, we can add `message.channels` to get the permissions to do so.

![After adding the permission and the app is saved, it should look like this](./added_channels_read.png)

If you look through the code that we now have in the `index.js` file, you'll see that we're listening for `messages`:

```javascript
slackEvents.on('message', (event) => {
  console.log(`Received a message event: user ${event.user} in channel ${event.channel} says ${event.text}`);
});
```

I can hear you asking "But here we're requesting `message.channels`, how do we know that those two match each other?"

> You can actually check the event `type` from [the API reference documentation](https://api.slack.com/events/message.channels) to see that the `type`s match up.

Author: `type` is a property and as such I like to leave in code fense. Leaving as-is

![The settings page with a button labeled "Reveal Config Vars"](./heroku_hidden_config_vars.png)

Upon opening the tab, you should see a button labeled "Reveal Config Vars". Press the button and copy your environmental variables from your `.env` file into the fields available.

![All of the .env file variables with the same name saved into Heroku](./heroku_config_vars.png)

Now that we have that, we can go back to our instructions that were on the main dashboard. Let's open up the same folder we have our `package.json` and `index.js` in, and install the Heroku CLI:

```
npm i -g heroku
```

> It's worth noting that Heroku official suggests using [an alternative installation method for the CLI](https://devcenter.heroku.com/articles/heroku-cli) due to Node.JS incompatibilities, but I've faced no such issues with my (admittedly limited) usage

Once this is done, we can:

- Sign into our account using `heroku login`

- Initialize a git repo into this folder `git init`

- Add Heroku as a remote place to deploy to. We should be able to see our Git URL from the settings page, so for example I would run:

  ```
  git remote add heroku https://git.heroku.com/points-r-us.git
  ```

> 

Author: ```suggestion
```

Now that we have Heroku setup, we're able to `git push heroku master` to have Heroku deploy our `npm start` script. This means that anything we put in our `package.json`'s `start` `script` property, then commit and push, will then be ran on our server. As such, the first thing we need to do is verify that we own that subdomain for Slack to send events to.

While our `package.json` might have looked like this before:

```json
"scripts": {
  "start:dev": "env-cmd node ./index.js",
  "start": "node ./index.js",
},
```

We'll want to update it so that the `start` command uses the signing secret from our server environmental variables to verify:

```json
"verify": "slack-verify --secret $SLACK_SIGNING_SECRET --port=$PORT",
"start": "npm run verify",
```

We need to allow Heroku to dictate the port to host our verification command as well, to get past their firewall they automatically route to the app's subdomain; hence the `--port` attribute. 

After making this change, we'll run:

- `git commit -m "Enforced verification"`
- `git push heroku master`

And watch as our app gets deployed

> ![The app being deployed during the `git push`](./heroku_initial_deploy.png)

Author: Good catch, thanks. It's not huge but I might as well blur it
